File: 1388-greatest-sum-divisible-by-three/greatest-sum-divisible-by-three.java
class Solution {
    public int maxSumDivThree(int[] nums) {
        int sum = 0;
        ArrayList<Integer> r1 = new ArrayList<>(), r2 = new ArrayList<>();

        for (int x : nums) {
            sum += x;
            if (x % 3 == 1) r1.add(x);
            else if (x % 3 == 2) r2.add(x);
        }
        if (sum % 3 == 0) return sum;

        Collections.sort(r1);
        Collections.sort(r2);

        int rem = sum % 3;
        int remv = Integer.MAX_VALUE;

        if (rem == 1) {
            if (r1.size() >= 1) remv = Math.min(remv, r1.get(0));
            if (r2.size() >= 2) remv = Math.min(remv, r2.get(0) + r2.get(1));
        } else {
            if (r2.size() >= 1) remv = Math.min(remv, r2.get(0));
            if (r1.size() >= 2) remv = Math.min(remv, r1.get(0) + r1.get(1));
        }
        return remv == Integer.MAX_VALUE ? 0 : sum - remv;
    }
}

File: 1388-greatest-sum-divisible-by-three/greatest-sum-divisible-by-three.cpp
class Solution {
public:
    int maxSumDivThree(vector<int>& nums) {
        int sum = 0;
        vector<int> remain1, remain2;

        for (int num : nums) {
            sum += num;
            if (num % 3 == 1) remain1.push_back(num);
            else if (num % 3 == 2) remain2.push_back(num);
        }

        if (sum % 3 == 0) return sum;

        sort(begin(remain1), end(remain1));
        sort(begin(remain2), end(remain2));

        int rem = sum % 3;
        int removeVal = INT_MAX;

        // Case when sum % 3 == 1
        if (rem == 1) {
            if (remain1.size() >= 1) 
                removeVal = min(removeVal, remain1[0]);
            if (remain2.size() >= 2) 
                removeVal = min(removeVal, remain2[0] + remain2[1]);
        }

        // Case when sum % 3 == 2
        else {  
            if (remain2.size() >= 1) 
                removeVal = min(removeVal, remain2[0]);
            if (remain1.size() >= 2) 
                removeVal = min(removeVal, remain1[0] + remain1[1]);
        }

        return (removeVal == INT_MAX) ? 0 : sum - removeVal;
    }
};
